### PR TITLE
pkg/errors: add multi error

### PR DESCRIPTION
**Pull request by sttts**
### Description of your changes

Follow-up of https://github.com/crossplane/crossplane/pull/4311.

Add `errors.Join(errs ...error)` and `errors.MultiError` interface. It is based on kube apimachinery
`NewAggregate` as the output is not multi-line like with Golang's `errors.Join`.

Fixes #

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Review by phisco — APPROVED** [review]
LGTM, if down the line upstream `errors.Join` will allow setting a custom separator instead of forcing `\n` we could easily switch to that.

File: pkg/errors/errors.go [hunk 1]
```diff
@@ -124,3 +126,34 @@ func Cause(err error) error {
...
 	return err
 }
+
+// MultiError is an error that wraps multiple errors.
+type MultiError interface {
+	error
+	Unwrap() []error
+}
+
+// Join returns an error that wraps the given errors.
```
Comment (negz) [inline pkg/errors/errors.go:136]: > aggregate errors if there is clear use for the user, but not in general, i.e. fail fast is the default.

I strongly agree with this guidance you added in https://github.com/crossplane/crossplane/pull/4311. My concern with wrapping errors is that folks will reach for the tool too often. What do you think about including this recommendation here in the docstring?

File: pkg/errors/errors.go [hunk 2]
```diff
@@ -22,6 +22,8 @@ package errors
 import (
 	"errors"
 	"fmt"
+
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
```
Comment (negz) [inline pkg/errors/errors.go:26]: Out of scope for this PR, but I think we use a few other things from `kerrors` pretty frequently that may be worth aliasing in this package. Can't remember what they are right now. :)

**Review by negz — APPROVED** [review]

File: pkg/errors/errors.go (hunk 2 above)
Comment (sttts) [reply pkg/errors/errors.go:26]: IsNotFound and the like?

File: pkg/errors/errors.go (hunk 1 above)
Comment (sttts) [reply pkg/errors/errors.go]: done

File: pkg/errors/errors.go [hunk 3]
```diff
@@ -124,3 +126,38 @@ func Cause(err error) error {
...
+// Join returns an error that wraps the given errors. Any nil error values are
+// discarded. Join returns nil if errs contains no non-nil values. The error
+// formats as the concatenation of the strings obtained by calling the Error
+// method of each element of errs and formatting like:
+//
+//	[first error, second error, third error]
+//
+// Note: aggregating errors should not be the default. Usually, return only the
+//
+//	first error, and only aggregate if there is clear value to the user.
```
Comment (negz) [inline pkg/errors/errors.go:145]: Looks like the formatting got a little messed up here?

File: pkg/errors/errors.go (hunk 2 above)
Comment (negz) [reply pkg/errors/errors.go:26]: Yeah. No need to add them now though, can follow up later.

File: pkg/errors/errors.go (hunk 3 above)
Comment (sttts) [reply pkg/errors/errors.go:145]: indeed

Comment (sttts) [reply pkg/errors/errors.go:145]: fixed

**Comment by phisco** [comment]
@sttts the linter is complaining because of the imports ordering, you can fix it automatically by running `GO_LINT_ARGS=" --fix " make reviewable` (which under the hood runs `golangci-lint run --fix` with the right version of golangci-lint)

**Pull request merged**